rework plugin structure and implement frontside blogpost (#46883)
sourcegraph/sourcegraph: client/backstage-common/README.md

- common-plugin: a plugin that is more a library, with common functionality you want to share across plugins. Unlike the other plugin types, this plugin cannot be run on it's own.

### How to generate plugin scaffolding.

1. Make sure you're in the Backstage repo generated with `@backstage/create-app` (Not a hard requirement, but the rest of the points assume this directory).
2. Run `backstage-cli new`, where a mini wizard will ask you questions about your plugin.
3. Once the command completes, the generated plugin scaffolding can be found under `plugins/`.

The plugin scaffolding now has been generated but the scaffolding assumes the plugin will stay in the Backstage repo which it won't. To make the plugin 'workable' outside of the Backstage repository we have make some additional changes.

1. Add the following `tsconfig.json` to the plugin directory.

```json
{
  "extends": "@backstage/cli/config/tsconfig.json",
  "include": ["src"],
  "compilerOptions": {
    "outDir": "dist-types",
    "rootDir": ".",
    "jsx": "react-jsx",
    "useUnknownInCatchVariables": false
  }
}
```

2. Take note of the plugin directory structure.

> ```bash

Reviewer: Something I learned recently is that you can do ` ```console` for console output blocks!

│   └── setupTests.ts
└── tsconfig.json

17 directories, 27 files
```

3. We need to edit the `package.json` to make it compatible to be imported from outside. Edit the `package.json` to look like the following.

```json
{
  "name": "backstage-plugin-sourcegraph-common",
  "description": "Common functionalities for the Sourcegraph plugin",
  "version": "0.1.0",
  "main": "dist/index.cjs.js",
  "types": "dist/index.d.ts",
  "license": "Apache-2.0",
  "private": true,
  "publishConfig": {
    "access": "public",
    "main": "dist/index.cjs.js",
    "module": "dist/index.esm.js",
    "types": "dist/index.d.ts"
  },
  "backstage": {
    "role": "common-library"
  },
  "scripts": {
    "build": "backstage-cli package build",
    "lint": "backstage-cli package lint",
    "test": "backstage-cli package test",
    "clean": "backstage-cli package clean",
    "prepack": "backstage-cli package prepack",
    "postpack": "backstage-cli package postpack"
  },
  "devDependencies": {
    "@backstage/cli": "^0.22.1"
  },
  "dependencies": {
    "@backstage/catalog-model": "^1.1.5",
    "@backstage/config": "^1.0.6",
    "@backstage/plugin-catalog-backend": "^1.7.1"
  },
  "files": ["dist"]
}
```

> The two important edits are changing the values for the properties `main`, `type` to point to the `dist` directory. 4. We can now copy the directory to wherever we want, like the Sourcgraph repo `cp plugins/backstage-plugin-test ~/sourcegraph/client/backstage/test`. 5. Depending on the type of plugin that was generated, a dependency entry was added to either the `packages/app/package.json  or `packages/backend/packages.json`. Now the that plugin has been 'moved' we don't want Backstage to still refer to these locations so remove the plugin with `yarn workspace backend remove <name>`or`yarn workspace app remove <name>`.

Reviewer: ```suggestion
The two important edits are changing the values for the properties `main`, `type` to point to the `dist` directory.

4. We can now copy the directory to wherever we want, like the Sourcgraph repo `cp plugins/backstage-plugin-test ~/sourcegraph/client/backstage/test`. 

5. Depending on the type of plugin that was generated, a dependency entry was added to either the `packages/app/package.json or `packages/backend/packages.json`. Now that the plugin has been 'moved' we don't want Backstage to still refer to these locations, so remove the plugin with `yarn workspace backend remove <name>`or`yarn workspace app remove <name>`.
```

### Run the plugin with Backstage / Local Development

Since we copied the plugin to a different directory Backstage we need to tell Backstage where to find the plugin.

> 1. Make sure the plugin is build. Note we use pnpm here and not yarn, that is because Sourcegraph uses pnpm. Both yarn and pnpm work with the command defined in the `package.json` which just executes `backstage-cli` with some args.

Reviewer: ```suggestion
1. Make sure the plugin is built. Note that we use pnpm here and not yarn, which is because Sourcegraph uses pnpm. Both yarn and pnpm work with the command defined in the `package.json` which just executes `backstage-cli` with some args.
```

> ```bash

Reviewer: ````suggestion
```console
````

$ cd sourcegraph/client/backstage/plugin
$ pnpm tsc # this generates required files in dist-types
$ pnpm build
```

> 2. There should be a dist directory now in `sourcegraphh/client/backstage/plugin`.

Reviewer: ```suggestion
2. There should now be a dist directory in `sourcegraph/client/backstage/plugin`.
```

> 3. We're now ready to show Backstage where the plugin is. Move the Backstage app directory and from the route execute

Reviewer: Move the app directory where?

   For a plugin that integrates with the backend:

> ```json

Reviewer: ````suggestion
```console
````

$ yarn workspace backend add link:~/sourcegraph/client/backstage/plugin
```

For a plugin that integrates with the frontend:

> ```json

Reviewer: ````suggestion
```console
````

6. We can now copy the directory to wherever we want, like the Sourcgraph repo `cp plugins/backstage-plugin-test ~/sourcegraph/client/backstage/test`.

7. Depending on the type of plugin that was generated, a dependency entry was added to either the `packages/app/package.json or `packages/backend/packages.json`. Now that the plugin has been 'moved' we don't want Backstage to still refer to these locations, so remove the plugin with `yarn workspace backend remove <name>`or`yarn workspace app remove <name>`.

### Run the plugin with Backstage / Local Development

Since we copied the plugin to a different directory we need to tell Backstage where to find the plugin.

1. Make sure the plugin is built. Note that we use pnpm here and not yarn, which is because Sourcegraph uses pnpm. Both yarn and pnpm work with the command defined in the `package.json` which just executes `backstage-cli` with some args.

```console
$ cd sourcegraph/client/backstage/plugin
$ pnpm es
```

2. There should now be a dist directory in `sourcegraph/client/backstage/plugin`.

> 3. We're now ready to show Backstage where the plugin is. Move the Backstage app directory and from the route execute

Author: ```suggestion
3. We're now ready to show Backstage where the plugin is. Move to the Backstage root directory and execute:
```

   For a plugin that integrates with the backend:

```console
$ yarn workspace backend add link:~/sourcegraph/client/backstage/plugin/dist
```

For a plugin that integrates with the frontend:

```console
$ yarn workspace app add link:~/sourcegraph/client/backstage/plugin/dist
```

> 4. You should be able to start the Backstage app now with `yarn dev`

Author: Changed the name
